File: progression/niveaux.py
"""Déduction du niveau d'un exercice à partir de l'historique.

Le niveau n'est jamais stocké : il est recalculé depuis les séances déjà
enregistrées, selon une règle unique — **le plus haut palier jamais validé**.
Cette règle est idempotente (recalculable de zéro à tout moment), insensible à
l'ordre des séances, et applique d'elle-même la règle « pas de recul ».

Le jugement se fait **par exercice**, pas par séance : une séance abandonnée
après avoir bouclé ses quatre séries de développé couché prouve tout de même
quelque chose. Le statut `abandoned` est donc volontairement ignoré ici, à la
différence de `historique.database.statistiques_exercices`.
"""

from historique.database import recuperer_historique
from progression.paliers import (
    UNITE_REPETITIONS,
    UNITE_SECONDES,
    est_suivi_par_le_moteur,
    exercices_suivis,
    niveau_pour,
    palier,
    unite,
)
from session.circuit import (
    MODE_AMRAP,
    MODE_CHRONO,
    MODE_MAINTIEN,
    MODE_REPETITIONS,
)

#: Ce qu'un mode mesure. Un exercice enregistré en `maintien` ne peut pas
#: valider un barème en répétitions : les deux ne parlent pas de la même chose.
UNITE_PAR_MODE = {
    MODE_REPETITIONS: UNITE_REPETITIONS,
    MODE_AMRAP: UNITE_REPETITIONS,
    MODE_MAINTIEN: UNITE_SECONDES,
    MODE_CHRONO: UNITE_SECONDES,
}


def performance_realisee(exercice, unite_attendue):
    """Réduit un exercice de l'historique à ce qu'il *prouve* : (poids, séries, cible).

    `exercice` est une entrée de `recuperer_historique()` :
        {"nom", "mode", "poids", "series", "repetitions", "duree",
         "series_detaillees": [
             {"serie", "repetitions", "poids", "duree", "completee"}, ...
         ]}

    `unite_attendue` vaut UNITE_REPETITIONS ou UNITE_SECONDES et indique quel
    champ d'une série porte la performance (`repetitions` ou `duree`).

    Retourne un triplet `(poids, nombre_de_series, cible)` exploitable par
    `niveau_pour`, ou None si la ligne ne prouve rien.

    Deux règles gouvernent la lecture :

    - seules les séries menées au bout comptent (`completee`) — une série
      interrompue par un abandon ou par la navigation web ne prouve rien ;
    - c'est le **maillon faible** qui commande, pas la moyenne ni le maximum.
      « Objectif atteint sur TOUTES les séries » se traduit par un minimum sur
      les séries retenues, pour la cible comme pour le poids : 12, 12, 10, 12
      répétitions prouve 10, pas 12.

    Une ligne sans série détaillée retourne None plutôt que de se rabattre sur
    les colonnes agrégées : `exercices.repetitions` est une *somme* sur toutes
    les séries, dont la performance de la plus faible ne se déduit pas.
    """
    champ = "duree" if unite_attendue == UNITE_SECONDES else "repetitions"

    series = [
        serie
        for serie in exercice.get("series_detaillees", [])
        if serie.get("completee")
    ]
    if not series:
        return None

    poids = min((serie.get("poids") or 0) for serie in series)
    cible = min((serie.get(champ) or 0) for serie in series)
    return poids, len(series), cible


def niveau_prouve_par(exercice):
    """Le plus haut niveau qu'un exercice de l'historique valide, ou None."""
    nom = exercice.get("nom")
    if not est_suivi_par_le_moteur(nom):
        return None

    unite_attendue = unite(nom)
    if UNITE_PAR_MODE.get(exercice.get("mode")) != unite_attendue:
        return None

    performance = performance_realisee(exercice, unite_attendue)
    if performance is None:
        return None

    poids, series, cible = performance
    return niveau_pour(nom, poids, series, cible)


def niveaux_par_exercice(seances=None):
    """Niveau actuel de chaque exercice suivi, avec la séance qui l'a établi.

    Un même exercice peut apparaître deux fois dans une séance (`jambes_abdos`
    enchaîne deux fois la planche latérale) : chaque ligne est jugée pour
    elle-même et c'est la meilleure qui l'emporte, jamais leur somme.
    """
    seances = recuperer_historique() if seances is None else seances
    niveaux = {}

    for seance in seances:
        for exercice in seance.get("exercices", []):
            niveau = niveau_prouve_par(exercice)
            if niveau is None:
                continue
            meilleur = niveaux.get(exercice["nom"])
            if meilleur is None or niveau > meilleur["niveau"]:
                niveaux[exercice["nom"]] = {
                    "niveau": niveau,
                    "palier": palier(exercice["nom"], niveau),
                    "seance_id": seance.get("id"),
                    "date": seance.get("date"),
                }

    return niveaux


def niveau_actuel(nom_exercice, seances=None):
    """Niveau d'un exercice, ou None s'il est hors barème ou non suivi."""
    entree = niveaux_par_exercice(seances).get(nom_exercice)
    return entree["niveau"] if entree else None


def etat_niveau(nom_exercice, seances=None, niveaux=None):
    """Tout ce qu'un écran a besoin de savoir sur le niveau d'un exercice.

    Trois situations distinctes, qu'un affichage ne doit pas confondre :
    `niveau` à None = **hors barème** (aucune performance n'atteint le premier
    palier, ce qui n'est pas « niveau 0 ») ; `maximum_atteint` = le barème est
    épuisé, il n'y a plus de palier au-dessus ; sinon `suivant` décrit le
    palier suivant.

    Volontairement sans notion de « pourcentage d'avancement » : un niveau dit
    où l'on en est, pas ce qu'il reste à faire. La part de barème parcourue
    n'aurait de sens que face à un objectif, ce qui relève de la couche
    programme.
    """
    if not est_suivi_par_le_moteur(nom_exercice):
        return None

    niveaux = niveaux_par_exercice(seances) if niveaux is None else niveaux
    entree = niveaux.get(nom_exercice)
    niveau = entree["niveau"] if entree else None
    suivant = palier(nom_exercice, (niveau or 0) + 1)

    return {
        "niveau": niveau,
        "actuel": palier(nom_exercice, niveau) if niveau else None,
        "suivant": suivant,
        "premier": palier(nom_exercice, 1),
        "maximum_atteint": niveau is not None and suivant is None,
        "date": entree["date"] if entree else None,
        "seance_id": entree["seance_id"] if entree else None,
    }


def etats_niveaux(seances=None):
    """État de niveau de tous les exercices suivis, en une seule lecture."""
    niveaux = niveaux_par_exercice(seances)
    return {
        nom: etat_niveau(nom, niveaux=niveaux) for nom in exercices_suivis()
    }

File: web/static/js/niveaux.js
// Jumeau de progression/niveaux.py — le niveau deduit de l'historique.
//
// **Le niveau est le plus haut palier jamais valide**, et il ne se stocke
// nulle part : il est recalcule a chaque lecture depuis l'historique. La
// regle est donc idempotente, insensible a l'ordre des seances comme a leur
// suppression, et elle applique d'elle-meme le « pas de recul ».
//
// A ne pas confondre avec l'**objectif** (`objectifs.js`), qui est ce que la
// prochaine seance demande : un plan, qui monte ou descend. Un objectif peut
// legitimement passer *sous* le niveau — c'est un delestage, pas une
// regression, et l'ecran des records continue d'afficher le niveau acquis.

import {
  MODE_AMRAP,
  MODE_CHRONO,
  MODE_MAINTIEN,
  MODE_REPETITIONS,
} from "./circuit.js";
import { UNITE_REPETITIONS, UNITE_SECONDES } from "./paliers.js";

//: Ce qu'un mode mesure. Un exercice enregistre en `maintien` ne peut pas
//: valider un bareme en repetitions : les deux ne parlent pas de la meme
//: chose.
export const UNITE_PAR_MODE = {
  [MODE_REPETITIONS]: UNITE_REPETITIONS,
  [MODE_AMRAP]: UNITE_REPETITIONS,
  [MODE_MAINTIEN]: UNITE_SECONDES,
  [MODE_CHRONO]: UNITE_SECONDES,
};

/**
 * Reduit un exercice de l'historique a ce qu'il *prouve* : [poids, series, cible].
 *
 * Deux regles gouvernent la lecture :
 *
 * - seules les series menees au bout comptent (`completee`) — une serie
 *   interrompue par un abandon ou par la navigation ne prouve rien ;
 * - c'est le **maillon faible** qui commande, pas la moyenne ni le maximum.
 *   « Objectif atteint sur TOUTES les series » se traduit par un minimum sur
 *   les series retenues, pour la cible comme pour le poids : 12, 12, 10, 12
 *   repetitions prouve 10, pas 12.
 *
 * Une ligne sans serie detaillee rend null plutot que de se rabattre sur les
 * colonnes agregees : `repetitions` y est une *somme* sur toutes les series,
 * dont la performance de la plus faible ne se deduit pas.
 */
export function performance_realisee(exercice, unite_attendue) {
  const champ = unite_attendue === UNITE_SECONDES ? "duree" : "repetitions";

  const series = (exercice.series_detaillees ?? []).filter((s) => s.completee);
  if (!series.length) return null;

  const poids = Math.min(...series.map((s) => s.poids || 0));
  const cible = Math.min(...series.map((s) => s[champ] || 0));
  return [poids, series.length, cible];
}

export class Niveaux {
  constructor(baremes) {
    this.baremes = baremes;
  }

  /** Le plus haut niveau qu'un exercice de l'historique valide, ou null. */
  niveau_prouve_par(exercice) {
    const nom = exercice.nom;
    if (!this.baremes.est_suivi_par_le_moteur(nom)) return null;

    const unite_attendue = this.baremes.unite(nom);
    if (UNITE_PAR_MODE[exercice.mode] !== unite_attendue) return null;

    const performance = performance_realisee(exercice, unite_attendue);
    if (performance === null) return null;

    const [poids, series, cible] = performance;
    return this.baremes.niveau_pour(nom, poids, series, cible);
  }

  /**
   * Niveau actuel de chaque exercice suivi, avec ce qui l'a etabli.
   *
   * Un meme exercice peut apparaitre deux fois dans une seance : chaque ligne
   * est jugee pour elle-meme et c'est la meilleure qui l'emporte, jamais leur
   * somme.
   *
   * Un **ancrage** recale le niveau d'un exercice que l'historique ne peut pas
   * prouver. Il agit de deux facons a la fois : il fait table rase de
   * l'historique anterieur, et il sert de plancher. C'est cette table rase qui
   * lui permet de corriger un niveau vers le bas — un simple plancher ne
   * saurait que le relever.
   */
  niveaux_par_exercice(seances, ancrages = {}) {
    const niveaux = {};

    for (const seance of seances) {
      for (const exercice of seance.exercices ?? []) {
        const nom = exercice.nom;
        const ancrage = ancrages[nom];
        // Anterieur a l'ancrage : ne compte plus.
        if (ancrage && (seance.id || 0) <= ancrage.apres_seance_id) continue;
        const niveau = this.niveau_prouve_par(exercice);
        if (niveau === null) continue;
        const meilleur = niveaux[nom];
        if (meilleur === undefined || niveau > meilleur.niveau) {
          niveaux[nom] = {
            niveau,
            palier: this.baremes.palier(nom, niveau),
            seance_id: seance.id ?? null,
            date: seance.date ?? null,
            ancre: false,
          };
        }
      }
    }

    for (const [nom, ancrage] of Object.entries(ancrages)) {
      if (!this.baremes.est_suivi_par_le_moteur(nom)) continue;
      const meilleur = niveaux[nom];
      if (meilleur === undefined || ancrage.niveau > meilleur.niveau) {
        niveaux[nom] = {
          niveau: ancrage.niveau,
          palier: this.baremes.palier(nom, ancrage.niveau),
          seance_id: null,
          date: ancrage.date,
          ancre: true,
        };
      }
    }

    return niveaux;
  }

  niveau_actuel(nom, seances, ancrages = {}) {
    const entree = this.niveaux_par_exercice(seances, ancrages)[nom];
    return entree ? entree.niveau : null;
  }

  /**
   * Tout ce qu'un ecran a besoin de savoir sur le niveau d'un exercice.
   *
   * **Trois situations distinctes, qu'un affichage ne doit pas confondre** :
   * `niveau` a null = hors bareme (aucune performance n'atteint le premier
   * palier, ce qui n'est pas « niveau 0 ») ; `maximum_atteint` = le bareme est
   * epuise ; sinon `suivant` decrit le palier suivant.
   *
   * Volontairement sans « pourcentage d'avancement » : un niveau dit ou l'on
   * en est, pas ce qu'il reste a faire. La part de bareme parcourue n'aurait
   * de sens que face a un objectif, ce qui releve de la couche programme.
   */
  etat_niveau(nom, niveaux) {
    if (!this.baremes.est_suivi_par_le_moteur(nom)) return null;

    const entree = niveaux[nom];
    const niveau = entree ? entree.niveau : null;
    const suivant = this.baremes.palier(nom, (niveau || 0) + 1);

    return {
      niveau,
      actuel: niveau ? this.baremes.palier(nom, niveau) : null,
      suivant,
      premier: this.baremes.palier(nom, 1),
      maximum_atteint: niveau !== null && suivant === null,
      date: entree ? entree.date : null,
      seance_id: entree ? entree.seance_id : null,
      ancre: Boolean(entree && entree.ancre),
    };
  }

  /**
   * Pour chaque seance, les exercices dont le niveau vient de monter grace a elle.
   *
   * Rejoue l'historique dans l'ordre **chronologique** (`recuperer_historique`
   * le rend le plus recent en tete) pour reperer l'instant precis ou chaque
   * niveau apparait. Meme regle de non-recul que `niveaux_par_exercice` et
   * meme traitement des ancrages ; la boucle est volontairement dupliquee
   * plutot que factorisee, l'une accumulant un maximum, l'autre un evenement
   * par franchissement.
   */
  montees_de_niveau(seances, ancrages = {}) {
    const chronologique = [...seances].sort((a, b) => (a.id || 0) - (b.id || 0));

    const niveaux = {};
    const montees = {};
    for (const seance of chronologique) {
      for (const exercice of seance.exercices ?? []) {
        const nom = exercice.nom;
        const ancrage = ancrages[nom];
        if (ancrage && (seance.id || 0) <= ancrage.apres_seance_id) continue;
        const niveau = this.niveau_prouve_par(exercice);
        if (niveau === null) continue;
        const avant = niveaux[nom];
        if (avant === undefined || niveau > avant) {
          montees[seance.id] = montees[seance.id] ?? {};
          montees[seance.id][nom] = {
            depuis: avant ?? null,
            vers: niveau,
            palier: this.baremes.palier(nom, niveau),
          };
          niveaux[nom] = niveau;
        }
      }
    }

    return montees;
  }

  /** Etat de niveau de tous les exercices suivis, en une seule lecture. */
  etats_niveaux(seances, ancrages = {}) {
    const niveaux = this.niveaux_par_exercice(seances, ancrages);
    const etats = {};
    for (const nom of this.baremes.exercices_suivis()) {
      etats[nom] = this.etat_niveau(nom, niveaux);
    }
    return etats;
  }
}
